impl/maven-impl/src/main/java/org/apache/maven/impl/DefaultPluginXmlFactory.java

```diff
@@ -45,58 +48,71 @@
         nonNull(request, "request");
         Path path = request.getPath();
         URL url = request.getURL();
         Reader reader = request.getReader();
         InputStream inputStream = request.getInputStream();
-        if (path == null && url == null && reader == null && inputStream == null) {
+        if (inputStream == null && reader == null && path == null && url == null) {
             throw new IllegalArgumentException("path, url, reader or inputStream must be non null");
         }
+        return read(request.isAddDefaultEntities(), request.isStrict(), inputStream, reader, path, url);
+    }
+
+    private static PluginDescriptor read(
+            boolean addDefaultEntities, boolean strict, InputStream inputStream, Reader reader, Path path, URL url) {
         try {
             PluginDescriptorStaxReader xml = new PluginDescriptorStaxReader();
-            xml.setAddDefaultEntities(request.isAddDefaultEntities());
+            xml.setAddDefaultEntities(addDefaultEntities);
             if (inputStream != null) {
-                return xml.read(inputStream, request.isStrict());
+                return read(xml, inputStream, strict);
             } else if (reader != null) {
-                return xml.read(reader, request.isStrict());
+                return xml.read(reader, strict);
             } else if (path != null) {
                 try (InputStream is = Files.newInputStream(path)) {
-                    return xml.read(is, request.isStrict());
-                }
-            } else {
-                try (InputStream is = url.openStream()) {
-                    return xml.read(is, request.isStrict());
+                    return read(xml, is, strict);
                 }
             }
-        } catch (Exception e) {
+            try (InputStream is = url.openStream()) {
+                return read(xml, is, strict);
+            }
+        } catch (IOException | XMLStreamException e) {
             throw new XmlReaderException("Unable to read plugin: " + getMessage(e), getLocation(e), e);
         }
     }
 
+    private static PluginDescriptor read(PluginDescriptorStaxReader xml, InputStream is, boolean strict)
+            throws XMLStreamException {
+        return xml.read(is, strict);
+    }
+
     @Override
     public void write(XmlWriterRequest<PluginDescriptor> request) throws XmlWriterException {
         nonNull(request, "request");
-        PluginDescriptor content = nonNull(request.getContent(), "content");
         Path path = request.getPath();
         OutputStream outputStream = request.getOutputStream();
         Writer writer = request.getWriter();
         if (writer == null && outputStream == null && path == null) {
-            throw new IllegalArgumentException("writer, outputStream or path must be non null");
+            throw new IllegalArgumentException("writer, output stream, or path must be non null");
```

Comment:
I don't think so.  Outputstream is the field on the request.

Reply:
yes thanks.

dedicated in: https://github.com/apache/maven/pull/2316